File: 3_Terzo Anno/AA 2023/1_Programmazione Logica e Funzionale (6 CFU)/Progetto [22]/Haskell/squirrelWarnsdroff.hs
{- Programma Haskell per risolvere il problema del giro del cavallo usando 
   l'algoritmo di Warnsdorff-Squirrel. -}

import Data.List (sortBy)
{- La libreria Data.List e' necessaria per utilizzare la funzione sortBy 
   che ordina una lista in base a un criterio specifico. -}

import Data.Ord (comparing)
{- La libreria Data.Ord e' necessaria per utilizzare la funzione comparing 
   che crea un criterio di ordinamento basato su una funzione di proiezione. -}

import Data.Maybe (listToMaybe)
{- La libreria Data.Maybe e' utilizzata per la funzione listToMaybe che converte 
   una lista in un Maybe, prendendo il primo elemento della lista se esiste. -}

import Data.IORef (IORef, newIORef, readIORef, writeIORef)
{- La libreria Data.IORef e' utilizzata per creare e manipolare riferimenti 
   mutabili (IORef) necessari per gestire lo stato mutabile nel programma. -}

import Text.Read (readMaybe)
{- La libreria Text.Read e' necessaria per utilizzare la funzione readMaybe 
   che tenta di leggere un valore da una stringa e restituisce Maybe. -}

-- Coordinate del cavallo
type Posizione = (Int, Int)

-- Scacchiera che contiene le mosse che compie il cavallo
type Scacchiera = [[Int]]

-- Movimenti del cavallo
mosseCavallo :: [Posizione]
mosseCavallo = [(2, 1), (1, 2), (-1, 2), (-2, 1), 
                (-2, -1), (-1, -2), (1, -2), (2, -1)]

main :: IO ()
main = do
    dimensione <- leggiDimensioneScacchiera
    posizioneIniziale <- leggiPosizione dimensione
    putStrLn "Attendere la soluzione..."
    risultato <- risolviGiroCavallo dimensione posizioneIniziale
    case risultato of
        Just soluzione -> stampaScacchiera soluzione
        Nothing        -> putStrLn "Soluzione non trovata."

{- Funzione che legge la dimensione della scacchiera da tastiera 
   e verifica la validita'.
   - restituisce un intero che rappresenta la dimensione della scacchiera. -}
leggiDimensioneScacchiera :: IO Int
leggiDimensioneScacchiera = do
    putStrLn "Inserisci la dimensione della scacchiera (intero compreso tra 5 e 70):"
    datoIngresso <- getLine
    case readMaybe datoIngresso of
        Just dimensione
            | dimensione < 5 -> do
                putStrLn "Non esiste una soluzione per scacchiere "
                putStrLn "di dimensioni inferiori a 5."
                leggiDimensioneScacchiera
            | dimensione > 70 -> do
                putStrLn "Con questo algoritmo non e' possibile risolvere il problema"
                putStrLn "per scacchiere di dimensioni superiori a 70."
                leggiDimensioneScacchiera
            | otherwise -> return dimensione
        Nothing -> do
            putStrLn "Dimensione non valida. Inserisci un numero intero."
            leggiDimensioneScacchiera

-- Funzione che legge una posizione da tastiera (da 1 a N) e la converte 
-- in coordinate valide per l'algoritmo (da 0 a N-1).
leggiPosizione :: Int -> IO (Int, Int)
leggiPosizione dimensione = do
    putStrLn "Inserisci la posizione di partenza del cavaliere in formato (X,Y)"
    putStrLn $ "(X e Y interi compresi tra 1 e " ++ show dimensione ++ "):"
    datoIngresso <- getLine
    -- Usa reads per cercare di interpretare l'input come una tupla
    case reads datoIngresso :: [((Int, Int), String)] of
        [((x, y), "")] 
            | x > 0 && x <= dimensione && y > 0 && y <= dimensione -> 
                return (x - 1, y - 1)  -- Converte le coordinate da 1-based a 0-based
        _ -> erroreInput dimensione

-- Funzione di errore che ristampa il messaggio di input non valido
erroreInput :: Int -> IO (Int, Int)
erroreInput dimensione = do
    putStrLn "Dato non valido. Inserisci di nuovo (esempio: (3,3)):"
    leggiPosizione dimensione

{- Funzione che inizializza una scacchiera NxN a -1.
   - il primo argomento e' la dimensione della scacchiera. -}
inizializzaScacchiera :: Int -> Scacchiera
inizializzaScacchiera dimensione = replicate dimensione (replicate dimensione (-1))

{- Funzione che verifica se una posizione e' valida.
   - il primo argomento e' la dimensione della scacchiera.
   - il secondo argomento e' la scacchiera.
   - il terzo argomento e' la posizione da verificare. -}
mossaValida :: Int -> Scacchiera -> Posizione -> Bool
mossaValida dimensione scacchiera (x, y) =
    x >= 0 && x < dimensione &&
    y >= 0 && y < dimensione &&
    (scacchiera !! x !! y) == -1

{- Funzione che calcola il numero di mosse valide successive da una data posizione.
   - il primo argomento e' la dimensione della scacchiera.
   - il secondo argomento e' la scacchiera.
   - il terzo argomento e' la posizione da cui calcolare l'accessibilita'. -}
calcolaAccessibilita :: Int -> Scacchiera -> Posizione -> Int
calcolaAccessibilita dimensione scacchiera (x, y) =
    length $ filter (mossaValida dimensione scacchiera) 
                    [(x + dx, y + dy) | (dx, dy) <- mosseCavallo]

{- Funzione che ordina le mosse in base all'accessibilita'.
   - il primo argomento e' la dimensione della scacchiera.
   - il secondo argomento e' la scacchiera.
   - il terzo argomento e' la lista delle posizioni da ordinare. -}
ordinaMosse :: Int -> Scacchiera -> [Posizione] -> [(Int, Posizione)]
ordinaMosse dimensione scacchiera mosse =
    sortBy (comparing fst) $
    map (\pos -> (calcolaAccessibilita dimensione scacchiera pos, pos)) mosse

{- Funzione che aggiorna la scacchiera con la nuova mossa.
   - il primo argomento e' la scacchiera.
   - il secondo argomento e' la posizione della mossa.
   - il terzo argomento e' il numero della mossa corrente. -}
aggiornaScacchiera :: Scacchiera -> Posizione -> Int -> Scacchiera
aggiornaScacchiera scacchiera (x, y) mossa =
    take x scacchiera ++
    [take y (scacchiera !! x) ++ [mossa] ++ drop (y + 1) (scacchiera !! x)] ++
    drop (x + 1) scacchiera

{- Funzione che risolve il problema del giro del cavallo usando l'algoritmo 
   di Warnsdorff-Squirrel.
   - il primo argomento e' la dimensione della scacchiera.
   - il secondo argomento e' la posizione di partenza del cavaliere. -}
risolviGiroCavallo :: Int -> Posizione -> IO (Maybe Scacchiera)
risolviGiroCavallo dimensione partenza = do
    riferimentoScacchiera <- newIORef (dimensione * dimensione, 
                                       inizializzaScacchiera dimensione)
    algoritmoWarnsdorffSquirrel dimensione (inizializzaScacchiera dimensione) 
                                partenza 1 riferimentoScacchiera

{- Funzione che implementa l'algoritmo di Warnsdorff-Squirrel per risolvere il giro 
   del cavallo.
   - il primo argomento e' la dimensione della scacchiera.
   - il secondo argomento e' la scacchiera corrente.
   - il terzo argomento e' la posizione corrente del cavallo.
   - il quarto argomento e' il numero della mossa corrente.
   - il quinto argomento e' un riferimento IORef contenente lo stato delle 
     caselle mancanti e la scacchiera. -}
algoritmoWarnsdorffSquirrel :: Int -> Scacchiera -> Posizione 
                            -> Int -> IORef (Int, Scacchiera) -> IO (Maybe Scacchiera)
algoritmoWarnsdorffSquirrel dimensione scacchiera posizione mossa 
                            riferimentoScacchiera = do
    let scacchieraAggiornata = aggiornaScacchiera scacchiera posizione mossa
    let caselleMancanti = dimensione * dimensione - mossa
    writeIORef riferimentoScacchiera (caselleMancanti, scacchieraAggiornata)
    if mossa == dimensione * dimensione
        then return (Just scacchieraAggiornata)
        else do
            let prossimeMosse = filter (mossaValida dimensione scacchieraAggiornata) 
                                [(fst posizione + dx, snd posizione + dy) 
                                | (dx, dy) <- mosseCavallo]
            if null prossimeMosse
                then return Nothing
                else do
                    let mosseOrdinate = ordinaMosse dimensione scacchieraAggiornata 
                                        prossimeMosse
                    tentaMossa mosseOrdinate scacchieraAggiornata
  where
    tentaMossa [] _ = return Nothing
    tentaMossa ((_, prossimaPosizione):resto) scacchieraCorrente = do
        risultato <- algoritmoWarnsdorffSquirrel dimensione scacchieraCorrente 
                                                 prossimaPosizione (mossa + 1) 
                                                 riferimentoScacchiera
        case risultato of
            Just soluzione -> return (Just soluzione)
            Nothing        -> tentaMossa resto scacchieraCorrente

{- Funzione che stampa la scacchiera.
   - l'unico argomento e' la scacchiera da stampare. -}
stampaScacchiera :: Scacchiera -> IO ()
stampaScacchiera scacchiera = mapM_ (putStrLn . unwords . map (pad . show)) scacchiera
  where pad s = replicate (3 - length s) ' ' ++ s

File: 3_Terzo Anno/AA 2023/1_Programmazione Logica e Funzionale (6 CFU)/Progetto [22]/ProLog/squirrelWarnsdroff.pl
/* Programma Prolog per risolvere il problema del giro del cavallo usando 
   l'algoritmo di Warnsdorff-Squirrel con ricorsione in coda. */

main :-
    leggi_dimensione_scacchiera(Dimensione),
    leggi_posizione(Dimensione, InizioX, InizioY),
    write('Attendere la soluzione...'), nl,
    (   risolvi(Dimensione, (InizioX, InizioY), ScacchieraFinale)
    ->  stampa_scacchiera(ScacchieraFinale)
    ;   write('Soluzione non trovata.'), nl
    ).

/* Predicato che legge e valida la dimensione della scacchiera. 
   - Dimensione: dimensione valida della scacchiera. */
leggi_dimensione_scacchiera(Dimensione) :-
    write('Inserisci la dimensione della scacchiera (intero compreso tra 5 e 70): '),
    read(Ingresso),
    (   integer(Ingresso), Ingresso >= 5, Ingresso =< 70 ->
        Dimensione = Ingresso
    ;   (integer(Ingresso) ->
            write('Dimensione non valida. Deve essere compresa tra 5 e 70.')
        ;   write('Dato inserito non valido. Inserisci un numero intero.')
        ), nl, leggi_dimensione_scacchiera(Dimensione)
    ).

/* Predicato chiamato in caso di errore durante la lettura della posizione. */
fallimento :-
    write('Dato inserito non valido o errore di sintassi. Riprova.\n'),
    fail.

/* Predicato che legge e valida la posizione iniziale del cavallo. 
   - N: dimensione della scacchiera.
   - InizioX, InizioY: coordinate valide della posizione iniziale. */
leggi_posizione(N, InizioX, InizioY) :-
    repeat,
    write('Inserisci la posizione di partenza del cavallo (X,Y)'),
    format(' (X e Y interi compresi tra 1 e ~d): ', [N]),  
    catch(read_term(user_input, Ingresso, []), _, fallimento),
    (   analizza_ingresso(Ingresso, N, InizioX, InizioY)
    ->  !
    ;   fallimento).

/* Predicato che analizza il dato in ingresso e verifica che sia 
   nel formato corretto (X,Y). 
   - Ingresso: il dato da analizzare.
   - N: dimensione della scacchiera.
   - X, Y: coordinate valide. */
analizza_ingresso((X,Y), N, X1, Y1) :-
    integer(X), integer(Y), X > 0, X =< N, Y > 0, Y =< N,  
    X1 is X - 1,  % Convertiamo le coordinate da 1-based a 0-based
    Y1 is Y - 1.
analizza_ingresso(_, _, _, _) :-
    fail.

/* Predicato che risolve il problema del giro del cavallo.
   - Dimensione: dimensione della scacchiera.
   - PosizioneIniziale: posizione iniziale del cavallo (InizioX, InizioY).
   - ScacchieraFinale: la scacchiera finale con il percorso del cavallo. */
risolvi(Dimensione, PosizioneIniziale, ScacchieraFinale) :-
    inizializza_scacchiera(Dimensione, Scacchiera),
    algoritmo_warnsdorff(Dimensione, Scacchiera, PosizioneIniziale, 1, ScacchieraFinale).

/* Predicato che implementa l'algoritmo di Warnsdorff-Squirrel con ricorsione in coda.
   - Dimensione: dimensione della scacchiera.
   - Scacchiera: scacchiera corrente.
   - (X, Y): posizione corrente del cavallo.
   - Mossa: numero della mossa corrente.
   - ScacchieraFinale: la scacchiera finale con il percorso completato. */
algoritmo_warnsdorff(Dimensione, Scacchiera, (X, Y), Mossa, ScacchieraFinale) :-
    aggiorna_scacchiera(Scacchiera, (X, Y), Mossa, ScacchieraAggiornata),
    (   Mossa =:= Dimensione * Dimensione  % Caso base: tutte le celle sono state visitate
    ->  ScacchieraFinale = ScacchieraAggiornata
    ;   findall((NX, NY), 
                (mosse_cavallo((X, Y), (NX, NY)), 
                 mossa_valida(Dimensione, ScacchieraAggiornata, (NX, NY))
                ), 
                Mosse),
        ordina_mosse(Dimensione, ScacchieraAggiornata, Mosse, MosseOrdinate),
        member(ProssimaMossa, MosseOrdinate),
        algoritmo_warnsdorff(Dimensione, ScacchieraAggiornata, ProssimaMossa,
                             Mossa + 1, ScacchieraFinale)
    ).

/* Predicato che inizializza la scacchiera con valori di default (-1).
   - N: dimensione della scacchiera (NxN).
   - Scacchiera: la scacchiera vuota inizializzata. */
inizializza_scacchiera(N, Scacchiera) :-
    length(Scacchiera, N),  % Crea una lista di N righe
    maplist(crea_riga(N), Scacchiera).

/* Predicato che crea una riga della scacchiera con valori di default (-1).
   - N: lunghezza della riga.
   - Riga: la riga creata con valori di default (-1). */
crea_riga(N, Riga) :-
    length(Riga, N),  % Crea una riga di lunghezza N
    maplist(=( -1), Riga).  % Inizializza tutte le celle a -1

/* Predicato che controlla se una mossa e' valida (ossia se e' dentro i confini 
   della scacchiera e la cella non e' stata visitata).
   - N: dimensione della scacchiera.
   - Scacchiera: scacchiera corrente.
   - (X, Y): posizione della mossa da controllare. */
mossa_valida(N, Scacchiera, (X, Y)) :-
    X >= 0, X < N, Y >= 0, Y < N,
    nth0(X, Scacchiera, Riga),
    nth0(Y, Riga, -1).

/* Predicato che calcola l'accessibilita' di una posizione 
    (numero di mosse valide possibili).
   - N: dimensione della scacchiera.
   - Scacchiera: scacchiera corrente.
   - (X, Y): posizione corrente.
   - Grado-(X, Y): il grado di accessibilita' e la posizione. */
calcola_accessibilita_con_posizione(N, Scacchiera, (X, Y), Grado-(X, Y)) :-
    findall((NuovoX, NuovoY), 
            (mosse_cavallo((X, Y), (NuovoX, NuovoY)), 
             mossa_valida(N, Scacchiera, (NuovoX, NuovoY))), Mosse),
    length(Mosse, Grado).

/* Predicato che ordina le mosse in base all'accessibilita' (secondo la regola di Warnsdorff).
   - N: dimensione della scacchiera.
   - Scacchiera: scacchiera corrente.
   - Mosse: lista delle mosse possibili.
   - MosseOrdinate: lista delle mosse ordinate per accessibilita'. */
ordina_mosse(N, Scacchiera, Mosse, MosseOrdinate) :-
    maplist(calcola_accessibilita_con_posizione(N, Scacchiera), Mosse, MosseConAccessibilita),
    keysort(MosseConAccessibilita, MosseOrdinateConAccessibilita),
    estrai_valori(MosseOrdinateConAccessibilita, MosseOrdinate).

/* Predicato che estrae i valori da una lista di coppie chiave-valore.
   - Lista: lista di coppie chiave-valore.
   - Valori: lista dei soli valori. */
estrai_valori([], []).
estrai_valori([_-Val|Resto], [Val|ValoriResto]) :-
    estrai_valori(Resto, ValoriResto).

/* Predicato che aggiorna la scacchiera con una nuova mossa.
   - Scacchiera: scacchiera corrente.
   - (X, Y): posizione della nuova mossa.
   - Mossa: numero della mossa corrente.
   - NuovaScacchiera: scacchiera aggiornata. */
aggiorna_scacchiera(Scacchiera, (X, Y), Mossa, NuovaScacchiera) :-
    (X >= 0, Y >= 0 ->  
        nth0(X, Scacchiera, Riga),  
        aggiorna_lista(Riga, Y, Mossa, NuovaRiga),  
        aggiorna_lista(Scacchiera, X, NuovaRiga, NuovaScacchiera)
    ;   throw(error(domain_error(not_less_than_zero, (X, Y)), _))).

/* Predicato che aggiorna una lista con un nuovo elemento in una posizione specifica.
   - Lista: la lista corrente.
   - Indice: posizione dell'elemento da aggiornare.
   - Elem: il nuovo elemento.
   - NuovaLista: lista aggiornata. */
aggiorna_lista([_|Resto], 0, Elem, [Elem|Resto]) :- !.
aggiorna_lista([Testa|Resto], Indice, Elem, [Testa|NuovoResto]) :-
    (Indice > 0 ->  
        NuovoIndice is Indice - 1,
        aggiorna_lista(Resto, NuovoIndice, Elem, NuovoResto)
    ;   throw(error(domain_error(not_less_than_zero, Indice), _))).

/* Predicato che genera le mosse possibili del cavallo a partire da una posizione.
   - (X, Y): posizione corrente del cavallo.
   - (NuovoX, NuovoY): nuova posizione possibile del cavallo. */
mosse_cavallo((X, Y), (NuovoX, NuovoY)) :-
    member((DeltaX, DeltaY), [(2, 1), (1, 2), (-1, 2), (-2, 1),
                              (-2, -1), (-1, -2), (1, -2), (2, -1)]),
    NuovoX is X + DeltaX,
    NuovoY is Y + DeltaY,
    NuovoX >= 0,
    NuovoY >= 0.

/* Predicato che stampa la scacchiera. */
stampa_scacchiera(Scacchiera) :-
    stampa_righe(Scacchiera).

/* Predicato che stampa tutte le righe della scacchiera. */
stampa_righe([]).
stampa_righe([Riga|Resto]) :-
    stampa_riga(Riga),
    nl,
    stampa_righe(Resto).

/* Predicato che stampa la scacchiera con numeri formattati */
stampa_riga([]).
stampa_riga([Elemento|Resto]) :-
    format('%4.0F ', [Elemento]),  
    stampa_riga(Resto).
